hash_writer.go

```diff
@@ -33,157 +33,249 @@ type HashWriter interface {
 }
 
 func (hashWriter *hashWriter) WriteFloat(val float64) error {
-	hashWriter.ionWriter.WriteFloat(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.FloatType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteFloat(val)
 }
 
 func (hashWriter *hashWriter) WriteDecimal(val *ion.Decimal) error {
-	hashWriter.ionWriter.WriteDecimal(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.DecimalType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteDecimal(val)
 }
 
 func (hashWriter *hashWriter) WriteTimestamp(val time.Time) error {
-	hashWriter.ionWriter.WriteTimestamp(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.TimestampType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteTimestamp(val)
 }
 
 func (hashWriter *hashWriter) WriteSymbol(val string) error {
-	hashWriter.ionWriter.WriteSymbol(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.SymbolType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteSymbol(val)
 }
 
 func (hashWriter *hashWriter) WriteString(val string) error {
-	hashWriter.ionWriter.WriteString(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.StringType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteString(val)
 }
 
 func (hashWriter *hashWriter) WriteClob(val []byte) error {
-	hashWriter.ionWriter.WriteClob(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.ClobType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteClob(val)
 }
 
 func (hashWriter *hashWriter) WriteBlob(val []byte) error {
-	hashWriter.ionWriter.WriteBlob(val)
-	panic("implement me")
+	error := hashWriter.hashScalar(ion.BlobType, val)
+	if error != nil {
+		return error
+	}
+	return hashWriter.ionWriter.WriteBlob(val)
 }
 
 func (hashWriter *hashWriter) BeginList() error {
+	error := hashWriter.stepIn(ion.ListType)
+	if error != nil {
+		return error
+	}
+
 	hashWriter.ionWriter.BeginList()
 	panic("implement me")
 }
 
 func (hashWriter *hashWriter) EndList() error {
-	hashWriter.ionWriter.EndList()
-	panic("implement me")
+	error := hashWriter.hasher.stepOut()
+	if error != nil {
+		return error
+	}
+
+	return hashWriter.ionWriter.EndList()
 }
 
 func (hashWriter *hashWriter) BeginSexp() error {
+	error := hashWriter.stepIn(ion.SexpType)
+	if error != nil {
+		return error
+	}
+
 	hashWriter.ionWriter.BeginSexp()
 	panic("implement me")
 }
 
 func (hashWriter *hashWriter) EndSexp() error {
-	hashWriter.ionWriter.EndSexp()
-	panic("implement me")
+	error := hashWriter.hasher.stepOut()
+	if error != nil {
+		return error
+	}
+
+	return hashWriter.ionWriter.EndSexp()
 }
 
 func (hashWriter *hashWriter) BeginStruct() error {
+	error := hashWriter.stepIn(ion.StructType)
+	if error != nil {
+		return error
+	}
+
 	hashWriter.ionWriter.BeginStruct()
 	panic("implement me")
```

Comment:
And this.
Or is this intentional because there is still more implementation left?

Reply:
Sorry, good catch!